GLPT_Cocoa.inc

```diff
@@ -1,89 +1,390 @@
+const
+  NSOpenGLPFAOpenGLProfile = 99 { available in 10_7 };
+const
+  NSOpenGLProfileVersionLegacy = $1000 { available in 10_7 };
+  NSOpenGLProfileVersion3_2Core = $3200 { available in 10_7 };
+  NSOpenGLProfileVersion4_1Core = $4100 { available in 10_10 };
+var
+  attributes: array[0..4] of NSOpenGLPixelFormatAttribute;
+begin
+  attributes[0] := NSOpenGLPFAWindow;
+  attributes[1] := NSOpenGLPFADoubleBuffer;
+  //NSOpenGLPFAColorSize, 32
+  //NSOpenGLPFADepthSize, 32
+  // todo: invalid NSOpenGLPixelFormat with core profile 
+  attributes[2] := NSOpenGLPFAOpenGLProfile;
+  attributes[3] := NSOpenGLProfileVersionLegacy;
+  attributes[4] := 0;
+
+  result := NSOpenGLPixelFormat.alloc.initWithAttributes(@attributes).autorelease;
+  if result = nil then
+    Fatal('invalid NSOpenGLPixelFormat');
+end;
+
+function TOpenGLView.initWithFrame(frameRect: NSRect): id;
+begin
+  result := inherited initWithFrame(frameRect);
+  if result <> nil then
+    begin
+      NSNotificationCenter(NSNotificationCenter.defaultCenter).addObserver_selector_name_object(result, objcselector('frameChanged:'), NSViewGlobalFrameDidChangeNotification, nil);
+    end;
+end;
+
+{=============================================}
+{@! ___GLTP___ } 
+{=============================================}
+
 function Cocoa_GetTime: double;
 begin
-  {$note implement this}
+  result := NSDate.date.timeIntervalSince1970;
 end;
 
 function Cocoa_Terminate: boolean;
 begin
-  {$note implement this}
-
+  NSApp.terminate(nil);
   exit(True);
 end;
 
 procedure Cocoa_DestroyWindow(var win: pGLPTwindow);
 begin
-  {$note implement this}
+  win^.ref.close;
+  win^.ref := nil;
 end;
 
 function Cocoa_CreateWindow(win: pGLPTwindow; posx, posy, sizex, sizey: integer; title: PChar): boolean;
+var
+  window: TCocoaWindow;
+  openGLView: TOpenGLView;
 begin
-  {$note implement this}
+  window := TCocoaWindow.alloc.initWithContentRect_styleMask_backing_defer(NSMakeRect(posx, posy, sizex, sizey), NSTitledWindowMask + NSClosableWindowMask + NSMiniaturizableWindowMask + NSResizableWindowMask, NSBackingStoreBuffered, false);
+  window.setTitle(NSSTR(title));
+  window.ref := win;
+
+  openGLView := TOpenGLView.alloc.initWithFrame(window.contentView.bounds);
+  window.setContentView(openGLView);
+  openGLView.release;
+
+  // todo: how do we stop beeping on mouse down??
+  window.makeFirstResponder(openGLView);
+
+  window.makeKeyAndOrderFront(NSApplication.sharedApplication);
+
+  win^.context := openGLView.openGLContext;
+  win^.ref := window;
 
   exit(True);
 end;
 
 function Cocoa_MakeCurrent(win: pGLPTwindow): boolean;
 begin
-  {$note implement this}
+  win^.context.makeCurrentContext;
 
   exit(True);
 end;
 
 procedure Cocoa_SwapBuffers(win: pGLPTwindow);
 begin
-  {$note implement this}
+  win^.ref.contentView.display;
 end;
 
 procedure Cocoa_GetFrameBufferSize(win: pGLPTwindow; out width, height: integer);
 begin
-  {$note implement this}
+  width:= trunc(win^.ref.contentView.bounds.size.width);
```

Comment:
Same here, if mac supports floats for the framebuffer size, should we consider using float on all platforms?